// doublescore/dist/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var types = __importStar(require("./lib/types"));
var clone_1 = __importDefault(require("./lib/clone"));
var iterate_1 = __importDefault(require("./lib/iterate"));
var mixin_1 = __importDefault(require("./lib/mixin"));
var timer_1 = __importDefault(require("./lib/timer"));
function DoubleScore() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var iterateWrapper = function (iterator) {
        return iterate_1.default.apply(iterate_1.default, args.concat(iterator));
    };
    iterateWrapper.flatten = function () {
        return iterate_1.default.flatten.apply(iterate_1.default, args);
    };
    return {
        iterate: iterateWrapper,
        clone: function () {
            return clone_1.default(args[0]);
        },
        getType: function () {
            return types.getType(args[0]);
        },
        isArray: function () {
            return types.isArray(args[0]);
        },
        isNumber: function () {
            return types.isNumber(args[0]);
        },
        isObject: function () {
            return types.isObject(args[0]);
        },
        mixin: function () {
            var myArgs = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                myArgs[_i] = arguments[_i];
            }
            return mixin_1.default.apply({}, args.concat(myArgs));
        }
    };
}
DoubleScore.clone = clone_1.default;
DoubleScore.getType = types.getType;
DoubleScore.isArray = types.isArray;
DoubleScore.isNumber = types.isNumber;
DoubleScore.isObject = types.isObject;
DoubleScore.iterate = iterate_1.default;
DoubleScore.mixin = mixin_1.default;
DoubleScore.timer = timer_1.default;
module.exports = DoubleScore;
